Deduplicate and filter breakpoints and call inversion variants

cwhelan opened the pull request:
This PR adds:

- A tool to take contig alignments from the AlignContigsAndCallBreakpointsSpark tool, deduplicate and filter those which have the same breakpoint structure, and add the proper annotations to the resulting VCF of called inversion events.
 
- A similar tool that takes SAM records and does the same thing, looking at split/supplementary alignments.

For @SHuang-Broad to review. There's a lot of code here so we can go over some of it in person if you want.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -84,4 +76,24 @@ protected void runTool(final JavaSparkContext ctx) {
+     * @return A tuple with the breakpoint ID and string representation of an AssembledBreakpoint, or an empty iterator if the line did not have two comma-separated values
+     */
+    static AlignmentRegion parseAlignedAssembledContigLine(final String alignedAssemblyContigLine) {
+        final String[] split = alignedAssemblyContigLine.split("\t", -1);
```
Method doesn't seem to be used in this tool but rather `CallVariantsFromAlignedContigsSpark`. Would it make sense to move it there?

cwhelan replied:
I put it here because it parses the output of this tool, and I think it's nice to have a tool know how to parse its own output (eg if we change the output of this tool we can update this parsing logic at the same time). Does that seem reasonable?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
should there be a "+1"? Take example of two contigs A[10,100], B[80, 120], should the overlap be 100-80+1? or the interval is right-exclusive?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Does it make sense to change the name to `getTotoalHardClipedBases` or something similar?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Because this is a public method, it makes sense to document what is the expected order in the input array.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
It looks like that the 3-parameter ctor and the actor taking a read would put the cigar into a forward strand cigar. So I think it makes sense to change the field name here to something like `forwardStrandCigar`, because it is exposed to other classes in this package.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
should these be `final`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
I think it's worth it considering making this class extending htsjdk `Allele`.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssembledBreakpoint.java:
I think it's better to document the expected order of fields, considering it's public.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Did you mean to do this:
```java
return super.makeReadFilter().and(ReadFilterLibrary.MAPPED);
```

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
This `@VisibleForTesting` is redundant.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
I see, this is the "contigs bam". Please ignore the above comment.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -32,7 +30,7 @@
+    public static final int EXPECTED_CONTIGS_PER_ASSEMBLY = 15;
 
     @Argument(doc = "file for assembled breakpoint output", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
             fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
```
Should the doc be changed to "file for assembled putative breakpoint output". 
And as a general consideration, should we change form "breakpoint id" to "assembly id", to avoid confusion with the new class AssembledBreakpoint? (I probably started using this "breakpoint id" term and am the person to blame for this.)

cwhelan replied:
Yes, I've renamed `breakpointId` to `assemblyId` in many places. Good call.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
I think it makes sense to document that this "breakpointId" might by `null` (when constructed with GATKRead) so accessors should have a null-check, considering it has package-access level.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
And may change the name, `breakpointId` -&gt; `assemblyId`.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
To me, because the design now is such that `AssembledBreakpoint` is composed of `AlignmentRegion`'s, but `AssembledBreakpoint` doesn't have such a field as `breakpointId` whereas `AlignmentRegion`'s have, it makes sense to move the `breakpointId` field to `AssembledBreakpoint` class.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
To me, because the design now is such that `AssembledBreakpoint` is composed of `AlignmentRegion`'s, but `AssembledBreakpoint` doesn't have such a field as `breakpointId` whereas `AlignmentRegion`'s have, it makes sense to move the `breakpointId` field to `AssembledBreakpoint` class.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssembledBreakpoint.java:
`final`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
The summary could be changed to "Parse SAM records of local assembly re-aligned to ref and call SVs"

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Should the program group be Spark or SV?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
`URL` -&gt; `URI`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
 
-    public List<AssembledBreakpoint> alignContigs (final ContigsCollection contigsCollection) {
-        final List<AssembledBreakpoint> assembledBreakpoints = new ArrayList<>();
+    public List<AlignmentRegion> alignContigs(String breakpointId, final ContigsCollection contigsCollection) {
```
I believe this method, as the core method, needs a documentation for 
1. what it does (kind of obvious)
2. what type filters are applied to the alignment regions (no secondary?)
3. any specific order in the returned list (start position in contig?)

And the `String` parameter could be final.

cwhelan replied:
Added documentation.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+        return next.mqual < 60;
     }
 
     private Collector<AlignmentRegion, ?, ArrayList<AlignmentRegion>> arrayListCollector(final int size) {
```
Is it possible to move this to `SVUtils`?

cwhelan replied:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+    }
+
+    @VisibleForTesting
+    public static List<AssembledBreakpoint> getAssembledBreakpointsFromAlignmentRegions(final byte[] sequence, final List<AlignmentRegion> alignmentRegionList, final Integer minAlignLength) {
```
This method is only used in `CallVariantsFromAlignedContigsSpark`. Would it make sense to move it there?
And the next three utility methods seem to only serve this method, so they may be moved together with this, if you decide to move them.

SHuang-Broad replied:
And actually when these four methods are moved to `CallVariantsFromAlignedContigsSpark`, they could all become `private` instead of `public`.

cwhelan replied:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Why is it an `Iterable` of reads instead of just one GATKRead?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
I think this methods (or the tool itself) needs some documentation on what filters are applied.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+        final List<AssembledBreakpoint> results = new ArrayList<>(alignmentRegionList.size() - 1);
+        final Iterator<AlignmentRegion> iterator = alignmentRegionList.iterator();
+        final List<String> insertionAlignmentRegions = new ArrayList<>();
+        if ( iterator.hasNext() ) {
```
It seems that if the input `alignmentRegionList` is empty, an exception might be thrown three lines above (0 - 1) -&gt; -1, can't create a list of size -1..., so if this check is necessary, it would be better to use `Utils.validateArgs(!alignmentRegionList.isEmpty(), "alignment region list cannot be empty to make breakpoints out of it.")`

cwhelan replied:
I just added a check to return an empty list of results if the incoming list is empty.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+            while (treatAlignmentRegionAsInsertion(current) && iterator.hasNext()) {
+                current = iterator.next();
+            }
+            while ( iterator.hasNext() ) {
```
A very-unlikely edge case: right now if all `AlignmentRegion`'s in the input has mapping quality less than 60, then the while loop will fall through, and an empty list would be returned. Just in case.

cwhelan replied:
I think that's fine and the intended behavior.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+                        SequenceUtil.reverseComplement(homologyBytes, 0, homologyBytes.length);
+                    }
+                    homology = new String(homologyBytes);
+                }
```
this if block can be replaced by a refactor (personal taste--I tend to avoid big `if`, `while`, and `for` blocks. this and the next two make the enclosing `while` block a bit hard to read):

```java
    String homology = getHomology(current, previous, sequenceCopy);

    private static String getHomology(final AlignmentRegion current, final AlignmentRegion previous, final byte[] sequenceCopy) {
        String homology = "";
        if (previous.endInAssembledContig >= current.startInAssembledContig) {
            final byte[] homologyBytes = Arrays.copyOfRange(sequenceCopy, current.startInAssembledContig - 1, previous.endInAssembledContig);
            if (previous.referenceInterval.getStart() > current.referenceInterval.getStart()) {
                SequenceUtil.reverseComplement(homologyBytes, 0, homologyBytes.length);
            }
            homology = new String(homologyBytes);
        }
        return homology;
    }
```

cwhelan replied:
good point, I've extracted `getHomology` and `getInsertedSequence` methods

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,52 +51,106 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+                        SequenceUtil.reverseComplement(insertedSequenceBytes, 0, insertedSequenceBytes.length);
+                    }
+                    insertedSequence = new String(insertedSequenceBytes);
+                }
```
Similar comment

```java
    String insertedSequence = getInsertedSequence(current, previous, sequenceCopy);

    private static String getInsertedSequence(final AlignmentRegion current, final AlignmentRegion previous, final byte[] sequenceCopy) {
        String insertedSequence = "";
        if (previous.endInAssembledContig < current.startInAssembledContig - 1) {

            final int insertionStart;
            final int insertionEnd;

            insertionStart = previous.endInAssembledContig + 1;
            insertionEnd = current.startInAssembledContig - 1;

            final byte[] insertedSequenceBytes = Arrays.copyOfRange(sequenceCopy, insertionStart - 1, insertionEnd);
            if (previous.referenceInterval.getStart() > current.referenceInterval.getStart()) {
                SequenceUtil.reverseComplement(insertedSequenceBytes, 0, insertedSequenceBytes.length);
            }
            insertedSequence = new String(insertedSequenceBytes);
        }
        return insertedSequence;
    }
```

cwhelan replied:
done as mentioned above

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
I propose the following change to the code:

1. remove functions `is5To3Inversion` and `is5To3Inversion`;
2. To make things work, add the following to class `BreakpointAllele`
    ```java
public enum InversionType{
        INV_3_TO_5, INV_5_TO_3, INV_UNKNOWN
    }

    public InversionType getInversionType(){
        if(!fiveToThree && threeToFive){
            return InversionType.INV_3_TO_5;
        }else if(fiveToThree && !threeToFive){
            return InversionType.INV_5_TO_3;
        }
        return InversionType.INV_UNKNOWN;
    }
    ```
3. To get string representation of the inversion type, it would now become something like `breakpoingAllele.getInversionType().name()`.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -84,4 +76,24 @@ protected void runTool(final JavaSparkContext ctx) {
         assembledBreakpoints.saveAsTextFile(output);
     }
 
+    static JavaPairRDD<String, ContigsCollection> getContigsCollectionKeyedByBreakpointId(final JavaSparkContext ctx, final String inputPath) {
```
What about a documentation
"Load a single local assembly flat file and turn into {@link ContigsCollection}, keyed by assembly id."?

cwhelan replied:
Changed name to `loadContigsCollectionKeyedByAssemblyId` to make it clear that we are loading from disk, and added a comment.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -22,7 +20,7 @@
 @CommandLineProgramProperties(summary="Align assembled contigs to the reference and call breakpoints from them.",
```
My understanding of the code is that this tool no longer calls breakpoint, but delegates work to `CallVariantsFromAlignedContigsSpark`?

cwhelan replied:
yes, updated doc

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -22,7 +20,7 @@
         oneLineSummary="Align contigs and call breakpoints",
         programGroup = SparkProgramGroup.class)
-public class AlignContigsAndCallBreakpointsSpark extends GATKSparkTool {
+public class AlignAssembledContigsSpark extends GATKSparkTool {
```
Should command line tools that can be called by user be `final`?

SHuang-Broad replied:
If you think so, the other two (`CallVariantsFromAlignedContigsSpark` and `CallVariantsFromAlignedContigsSAMSpark`) should be marked accordingly as well.

cwhelan replied:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
To me this method is better named as `inversionBreakpointAlleleFilter`, because `AssembledBreakpoint`'s are not filtered, rather it is the `BreakpointAllele`'s (i.e. keys of input).

SHuang-Broad commented:
A may-be-too-early comment: the two tools `CallVariantsFromAlignedContigsSpark` and `CallVariantsFromAlignedContigsSAMSpark` share a significant part of workflow. Conceptually, what happens is:

1. turn `AlignmentRegion`'s (produced by another tool in difference formats that must be parsed) into `AssembledBreakpoint`'s;
2. extract `BreakpointAllele` from a `AssembledBreakpoint`, and group the `AssembledBreakpoint`'s by each of their `BreakpointAllele`;
3. filter `BreakpointAllele`'s;
4. for each `BreakpointAllele` that survived the filter, filter its supporting `AssembledBreakpoint`'s then create an `VariantContext`;
5. write VCF.

In both of the tools, these steps are accomplished in almost the same way, involving the following functions
```java
public static Iterable<AssembledBreakpoint> assembledBreakpointsFromAlignmentRegions(final byte[] contigSequence,
                                                                                     final Iterable<AlignmentRegion> alignmentRegionsIterable,
                                                                                     final Integer minAlignLength);

public static Tuple2<BreakpointAllele, Tuple2<Tuple2<String,String>, AssembledBreakpoint>> keyByBreakpointAllele(final Tuple2<Tuple2<String,String>, AssembledBreakpoint> breakpointIdAndAssembledBreakpoint);

public static Boolean inversionBreakpointFilter(final Tuple2<BreakpointAllele, Tuple2<Tuple2<String,String>,AssembledBreakpoint>> breakpointAlleleTuple2Tuple2);

public static VariantContext filterBreakpointsAndProduceVariants(final Tuple2<BreakpointAllele, Iterable<Tuple2<Tuple2<String,String>, AssembledBreakpoint>>> assembledBreakpointsPerAllele,
                                                                 final Broadcast<ReferenceMultiSource> broadcastReference);
```

The concept is general enough that a separate class (potentially having its own shallow class hierarchy for different types of SV's) is extracted and provide a single interface like `callVariants()`.
And if the afore-mentioned functions could be moved from `ContigAligner` to here, all of them can be packed in the extracted class/interface.
Then different callers for different SV types could just override (or whatever technique) when fit.

SHuang-Broad commented:
Done for the 1st round.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Good catch. Fixed and added a test.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
It's supposed to match the output of 'toString()'. Added documentation.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
Yes, done.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
I think that that `Allele` class has a lot of stuff that's not relevant to this class, and vice versa, so it might be more trouble than its worth to make one extend the other. For example, the htjdk class does not have any concept of genomic location, and is just concerned with the bases content of a ref or alternate allele. Perhaps this class shouldn't be called `BreakpointAllele`. I'll see if I can come up with different names that would make sense.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssembledBreakpoint.java:
Again this was designed to read the output of `toString()`. Added some documentation.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AssembledBreakpoint.java:
Actually just removed that try/catch since it was unnecessary.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Done, changed to "Parse a SAM file containing contigs or long reads aligned to the reference and call SVs"

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Done, and changed all the other tools that were still under plain old `SparkProgramGroup` to `StructuralVariationSparkProgramGroup`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
If you're talking about the second entry in the `alignmentsGroupedByName` tuple, it's because this includes supplementary alignments so you'll get multiple `GATKRead` objects for each read name.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
This issue with that is that we are storing `AlignmentRegion` objects and reading them in (in another tool) before converting them into `AssembledBreakpoint` objects. So I need to denormalize the ID for that reason.

cwhelan commented:
The naming of the different classes (`AlignmentRegion`, `AssembledBreakpoint`, `BreakpointAllele`) was very confusing especially with the id variable `breakpointId` in several classes. I've renamed `AssembledBreakpoint` to `BreakpointAlignment`, since I think that the main thing it's trying to capture is a split alignment that's indicating there might be a breakpoint at a given location. Then, the actual breakpoint is represented by `BreakpointAllele` which the information about the breakpoint junction, but not the alignment-related fields. Does that make more sense?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSAMSpark.java:
Added some documentation to the tool.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented:
In terms of the two tools, I don't think it's necessary at this point to make an inheritance structure. `CallVariantsFromAlignedContigsSAMSpark` is more of a one-off for dealing with de novo assembly files and I'm not sure if it will be supported long term. However, I did extract a `callVariantsFromAlignmentRegions` method in `CallVariantsFromAlignedContigsSpark` that `CallVariantsFromAlignedContigsSAMSpark` can use, which reduces code duplication a lot. There's not much left in `CallVariantsFromAlignedContigsSAMSpark` except for the logic to convert GATKReads into AlignmentRegions, which seems appropriate.

cwhelan commented:
I do like your suggested workflow in your long comment above, and I think it would be good to pursue something similar when we add additional SV types.

cwhelan commented:
Address PR comments to some extent. @SHuang-Broad take a look and let me know if things are more clear/better structured to you now. I think that renaming the `AssembledBreakpoint` class makes things at least a little bit less confusing.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointEvidence.java:
```diff
@@ -176,7 +176,7 @@ private static TemplateEnd findTemplateEnd( final GATKRead read ) {
             super(read, metadata, atStart ? read.getStart() : read.getEnd(), UNCERTAINTY);
             cigar = read.getCigar().toString();
-            if ( cigar == null ) throw new GATKException("Read has no cigar string.");
+            if ( cigar == null ) throw new GATKException("Read has no forwardStrandCigar string.");
```
I believe this change is an IDE auto-rename error, as this was a different part of the SV pipeline...

cwhelan replied:
Whoops, sorry. Thanks for catching that.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointEvidence.java:
```diff
@@ -236,7 +236,7 @@ public SplitRead read( final Kryo kryo, final Input input, final Class<SplitRead
             cigar = read.getCigar().toString();
-            if ( cigar == null ) throw new GATKException("Read has no cigar string.");
+            if ( cigar == null ) throw new GATKException("Read has no forwardStrandCigar string.");
         }
```
Again, IDE overkill...

cwhelan replied:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -81,7 +73,31 @@ protected void runTool(final JavaSparkContext ctx) {
+     * @param alignedAssembedContigLine An input line with the tab-separated fields of an alignment region
+     * @return A tuple with the breakpoint ID and string representation of an BreakpointAlignment, or an empty iterator if the line did not have two comma-separated values
+     */
+    static AlignmentRegion parseAlignedAssembledContigLine(final String alignedAssembedContigLine) {
```
typo: `alignedAssembedContigLine` -&gt; `alignedAssembledContigLine`

cwhelan replied:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
`protected` could be `private`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
It seems that this method could also be called within the public method (which itself could become package-access) `callVariantsFromAlignmentRegions` (and of course its name could be changed then to reflect the fact that it is also writing output).
It you decide this is worth it, then `writeVariants` could become private as well.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Could be `private`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Could be `private`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
As this becomes the workflow manager of the actual variant calling, it's worth it to document the input arguments, especially the second one (it's long and the first Tuple2 is not obviously telling its purpose):
`final JavaPairRDD<Tuple2<String, String>, Tuple2<Iterable<AlignmentRegion>, byte[]>> alignmentRegionsWithContigSequences`

SHuang-Broad commented:
Hi @cwhelan , just added some trivial comments, and which you could merge.

Yes, I agree the names are better, but I could be biased as I've spent sometime on this. Maybe @tedsharpe could provide some more spark on better names, if possible?

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVUtils.java:
```diff
@@ -124,4 +126,11 @@ public static int hashMapCapacity( final int nElements )
+    /**
+     * Provides a stream collector that will collect items into an array list with a given initial capacity.
+     */
+    static Collector<AlignmentRegion, ?, ArrayList<AlignmentRegion>> arrayListCollector(final int size) {
```
Maybe genericize like this:
    static <T> Collector<T, ?, ArrayList<T>> arrayListCollector(final int size) {

cwhelan replied:
Done

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -81,7 +73,31 @@ protected void runTool(final JavaSparkContext ctx) {
+     * should be the output of as RunSGAViaProcessBuilderOnSpark.
+     */
+    static JavaPairRDD<String, ContigsCollection> loadContigsCollectionKeyedByAssemblyId(final JavaSparkContext ctx, final String inputPath) {
+        final JavaRDD<String> inputAssemblies = ctx.textFile(inputPath).cache();
```
My understanding of Spark persistence is still pretty limited, so forgive me if this is stupid, but...
a) inputAssemblies is local to this method and referenced only once, so how does caching improve anything?
b) when does the memory get reclaimed and how?  seems like cache() ought to always be matched by unpersist().

cwhelan replied:
Thanks for catching that, the cache() was part of an earlier experiment that became obsolete when I reorganized the code. I've removed it.

But my understanding is that Spark will automatically unpersist RDDs from the cache as needed so you don't always need to call `unpersist()`. In the docs it says:

> Removing Data  
Spark automatically monitors cache usage on each node and drops out old data partitions in a     least-recently-used (LRU) fashion. If you would like to manually remove an RDD instead of   waiting for it to fall out of the cache, use the RDD.unpersist() method.

From http://spark.apache.org/docs/latest/programming-guide.html#rdd-persistence

Am I reading that correctly, or do you think unpersisting is necessary? (Just so I know what's going on going forward)

tedsharpe replied:
It can be evicted by other persistent data, but it's still occupying heap memory.  So that keeps us from re-using that space for user data or for RDD actions.  I could be all wet, but I think we probably need to unpersist.

cwhelan replied:
OK. Maybe we can experiment with this/learn more about it in the future. For now removed unnecessary call to `cache()`.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -81,7 +73,31 @@ protected void runTool(final JavaSparkContext ctx) {
+
+        final JavaPairRDD<String, String> contigCollectionByBreakpointId =
+                inputAssemblies
+                        .flatMapToPair(RunSGAViaProcessBuilderOnSpark::splitAssemblyLine);
```
It feels very strange to me to be calling static methods from one tool in a different tool.  (Examples: the method we're in: loadContigsCollectionKeyedByAssemblyId, or like this call to splitAssemblyLine in RunSGAViaProcessBuilderOnSpark.
It's probably just an atavism from days of C and C++ where if you tried that you'd end up with two main methods and linking would fail.
Also, I know that the organization of our jar is "a gigantic can of worms", where we have a driver program that can call any tool, and therefore all code with all its complex dependencies must always be available for loading.
But tangling the dependency tree in this way can't be good.
There.  I've had my say.  Feel free to ignore this comment.

cwhelan replied:
Hmm. I've just been trying to keep the code that reads and parses data near the code that writes that data, so that you only have to change it one place. Maybe it needs to be separated out into a persistence class though, since both you and @SHuang-Broad have complained about it.

tedsharpe replied:
What I've typically done in the past is to move a nested class to an independent, top-level class as soon as two programs need it.  So, by that rule, ContigsCollection should be in its own file.  And it could have some static methods for handling I/O.
I've never been completely happy about cluttering up a class like ContigsCollection with knowledge about I/O and external formats, but maybe that's the least evil.  I'd been putting that stuff in SVUtils, but I'm not really happy with that idea, either.  Maybe the right way to do it is to have another top-level class, ContigsCollectionIO.  Maybe we can chat about it and decide on a style, and stick with that.

cwhelan replied:
I've moved all of these parsing methods to `ContigsCollection` for now, and made it a top-level class. Does that look better to you?

tedsharpe replied:
Yes.  This makes me much happier.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignAssembledContigsSpark.java:
```diff
@@ -52,27 +50,21 @@ public boolean requiresReference() {
+        final JavaPairRDD<String, ContigsCollection> breakpointIdsToContigsCollection = loadContigsCollectionKeyedByAssemblyId(ctx, input).cache();
 
-        final long numInputPartitions = inputAssemblies.count();
+        final long numInputPartitions = breakpointIdsToContigsCollection.count();
```
This is misnamed:  it's the number of assemblies, nothing to do with partitions.

cwhelan replied:
Yes, thanks for noticing that. Fixed.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
While looking at the example VCF, I realized that this TOTAL_MAPPINGS field might be redundant, because this integer can always be inferred from other annotations.
I would guess that this is put here for convenience and may not lead to much spatial overhead.
Is this the case?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
Same comment as the one for TOTAL_MAPPINGS

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
mapqual might be a little more obvious

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
doesn't handle edge cases where cigarElements.size() == 0 (probably silly)
or cigarElements.size() == 1 (which is a little less silly, but not inconceivable).

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
misleading name.  nClipped?  result?

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
The cigar is arguably broken if it contains only clipping, but still, you might add:
if ( j < 0 || j >= cigar.getCigarElements().size() ) break;

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Appears to be unused.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
OK, now I'm getting really fussy, but maybe "NA" isn't the greatest choice here (and 2 lines above) since it could be mistaken for the two-base sequence 'N' (any base) + 'A' (adenosine).  Maybe "none", instead?

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
could use ==

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
Is the third possibility actually INV_NONE?  Or is there a legal ambiguous state?

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,56 +48,38 @@ public ContigAligner(final String referenceFilePath) throws IOException {
                 final byte[] sequence = contigInfo._2.toString().getBytes();
                 final AlnRgn[] alnRgns = bwaAlignSequence(bwaMem, contigId, sequence);
 
                 log.info("alnRgns : " + (alnRgns == null ? "null" : alnRgns.length));
```
did you intend to leave this in, or is this debug code?

cwhelan replied:
removed

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,56 +48,38 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+     * @return
+     */
+    public List<AlignmentRegion> alignContigs(final String assemblyId, final ContigsCollection contigsCollection) {
+        final List<AlignmentRegion> alignedContigs = new ArrayList<>();
```
can probably initialize this with capacity contigsCollection.getContents().size()

cwhelan replied:
done

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java:
```diff
@@ -54,56 +48,38 @@ public ContigAligner(final String referenceFilePath) throws IOException {
+                        .map(a -> new AlignmentRegion(assemblyId, contigId, a))
+                        .sorted(Comparator.comparing(a -> a.startInAssembledContig))
+                        .collect(SVUtils.arrayListCollector(alnRgns.length));
+                 alignedContigs.addAll(alignmentRegionList);
```
rather than collecting into a new list and then adding that list to the final list, your final pipeline step could just stuff the new ones directly into the final list:
.forEach(alignedContigs::add);

cwhelan replied:
Done, thanks, that's cleaner.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
I don't think naming the list adds any clarity.
`ComplicatedType varName = expression;
return varName;`
could just be
`return expression;`

tedsharpe commented:
That's all I have to offer.  Review complete.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Added some docs.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
Yeah, I was just thinking about convenience (and the ability to apply filters to the VCF with simple tools like GATK SelectVariants, VCFTools, etc.). I'm not too worried about the redundant data storage for now; our VCF files are really quite small.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
Same response as above.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
changed.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Added two edge case checks.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
You're right; changed name to `result`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Fussy is good. Changed.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
You're right, I didn't finish wiring this piece of data through. I'll fix that.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
Just trying to guard/point out a weird state. I can change it to INV_NONE, I guess, since it shouldn't really exist.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
Sorry to find this late, but this is probably better to be renamed to "ASSEMBLY_IDS" as well?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
And here as well.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GATKSVVCFHeaderLines.java:
Changed.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
This `+ 1` is worth investigating, as we talked about.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Removed the second +1

cwhelan commented:
@SHuang-Broad @tedsharpe I think I've addressed all the comments. Does this look good now?

I just ran through the entire pipeline and it looks like our raw number of inversions called has dropped a bit relative to the experimental code I was running before, but I'll save investigating that for another PR.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Maybe a better name is `getNumClippedBases()`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Would be better if this documentation is added to clear confusion:
`Assumes {@code region1} has a lower {@link AlignmentRegion#startInAssembledContig} than {@code region2}.`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Personally I still find this method's name, and the next, a little confusing.
I believe these two are trying to find reference locations of the 5'- and 3'-end breakpoints, so mentioning the AssembledContig is, IMO, not helping.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Totally a design choice, so you can ignore this comment: would it be better to move this method to the `BreakpointAllele` class and make it a factory method?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
What about replacing these two `boolean`'s with a 
`final InversionType inversionType;`, and initialize it with 
```java
if(!fiveToThree && threeToFive){
    return InversionType.INV_3_TO_5;
} else if(fiveToThree && !threeToFive) {
    return InversionType.INV_5_TO_3;
} else {
    return InversionType.INV_NONE;
}
```

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
Of course this would force the `hash()` function to be updated.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Another refactoring comment: could these be refactored into a call like 
`final JavaPairRDD<Tuple2<String, String>, Tuple2<Iterable<AlignmentRegion>, byte[]>> alignmentRegionsWithContigSequences = prepAlignmentRegionsForCalling(cox);`, where (IDE did the work)

```java
    private JavaPairRDD<Tuple2<String, String>, Tuple2<Iterable<AlignmentRegion>, byte[]>> prepAlignmentRegionsForCalling(JavaSparkContext ctx) {
        final JavaRDD<AlignmentRegion> inputAlignedContigs = ctx.textFile(inputAlignments).map(ContigsCollection::parseAlignedAssembledContigLine);

        final JavaPairRDD<Tuple2<String, String>, Iterable<AlignmentRegion>> alignmentRegionsKeyedByBreakpointAndContig = inputAlignedContigs.mapToPair(alignmentRegion -> new Tuple2<>(new Tuple2<>(alignmentRegion.assemblyId, alignmentRegion.contigId), alignmentRegion)).groupByKey();

        final JavaPairRDD<String, ContigsCollection> assemblyIdsToContigCollections = loadContigsCollectionKeyedByAssemblyId(ctx, inputAssemblies).cache();

        final JavaPairRDD<Tuple2<String, String>, byte[]> contigSequences = assemblyIdsToContigCollections.flatMapToPair(assemblyIdAndContigsCollection -> {
            final String assemblyId = assemblyIdAndContigsCollection._1;
            final ContigsCollection contigsCollection = assemblyIdAndContigsCollection._2;
            return contigsCollection.getContents().stream().map(pair -> new Tuple2<>(new Tuple2<>(assemblyId, pair._1.toString()), pair._2.toString().getBytes())).collect(Collectors.toList());
        });

        return alignmentRegionsKeyedByBreakpointAndContig.join(contigSequences);
    }
```

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignmentRegion.java:
Done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
Changed to `getLeftAlignedLeftBreakpointOnReference`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
That would be a valid choice but I'm not sure I see a compelling reason to switch it right now.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAllele.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented:
Made the changes you requested, @SHuang-Broad

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
IMO, something like `breakpointAlleleRepresentsInversion` is clearer because this method is testing if the breakpoint allele really represents an inversion, not trying to further filter the inversion breakpoint alleles.

A related minor issue is that the method should preferably take a single argument of `BreakpointAllele`, because the associated `BreakpointAlignment` is not used. Of course if this method is intended to be extended in the future, and the alignments may be needed, this should not be changed.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
changed this to a lambda that calls a method `BreakpointAllele.isInversion()` on the key of the pairRDD.

cwhelan commented:
@SHuang-Broad does this refactor to the inversion filter look better?

SHuang-Broad commented:
Yep, much better! Thanks!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
As we discussed, another possible change that easier for me to understand but not necessarily true for others:
the private method `getLeftAlignmentRegion()` is picking the AR (AlignmentRegion) with the lower reference-coordinate (starting position really, we could dream up a case that one is completely contained in the other but that's impossible for a good inversion contig, for now). So the three methods: `getLeftAlignmentRegion()`, `getLeftAlignedLeftBreakpointOnReference()`, `getLeftAlignedRightBreakpointOnReference()`, could be merged into a single method, something like:

```java
    /**
     * Computes the 5'- and 3'- end breakpoint locations, 
     * where 5'-end BP is left aligned on reference and 3'-end BP is right aligned when homology around breakpoints is present.
     */
    @VisibleForTesting
    Tuple2<SimpleInterval, SimpleInterval> inferBreakpointLocWithParsimony(){

        String fiveContig, threeContig;
        int fivePosition, threePosition;

        if (region1.referenceInterval.getStart() < region2.referenceInterval.getStart()) {
            fivePosition = region1.forwardStrand ? region1.referenceInterval.getEnd() - homology.length() : region1.referenceInterval.getStart();
            threePosition = region2.forwardStrand ? region2.referenceInterval.getStart() + homology.length() : region2.referenceInterval.getEnd();
            fiveContig = region1.referenceInterval.getContig();
            threeContig = region2.referenceInterval.getContig();
        } else {
            fivePosition = region2.forwardStrand ? region2.referenceInterval.getStart() : region2.referenceInterval.getEnd() - homology.length();
            threePosition = region1.forwardStrand ? region1.referenceInterval.getEnd() : region1.referenceInterval.getStart() + homology.length() ;
            fiveContig = region2.referenceInterval.getContig();
            threeContig = region1.referenceInterval.getContig();
        }

        return new Tuple2<>( new SimpleInterval(fiveContig, fivePosition, fivePosition),
                             new SimpleInterval(threeContig, threePosition, threePosition) );
    }
``` 

Reason for naming this function this way (Soohee helped me wit this): `BreakpointAlignment`, for an inversion, is either 5-to-3 or 3-to-5. (Another language system that I personally also use is 5-5 adjacency and 3-3 adjacency, to be inline with the "novel adjacency" language system.) The logic here is (IMO) to give the homologous sequence to 
1. the inferred 3'-end (therefore 5'-end BP is left-aligned) in the case of 5-to-3 type (or the 5-5 adjacency);
2. the inferred 5'-end (therefore 3'-end BP is right-aligned) in the case of 3-to-5 type (or the 3-3 adjacency);

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
This "inversions.vcf" file name could be extracted to a static string of the class.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
the list and of VCs code could be replace with a one liner:
```java
final List<VariantContext> sortedVariantsList = variantContextJavaRDD.collect().stream()
    .sorted((VariantContext v1, VariantContext v2) -> IntervalUtils.compareLocatables(v1, v2, referenceSequenceDictionary))
    .collect(Collectors.toList());
```

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
These two, `lowMqMappings` and `midMqMappings` are not used.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Could this method's name be changed? It doesn't appear to be filtering any breakpoints right now (although I believe in the future it could), or breakpoints are supposed to be filtered if they have zero `highMqMappings`?

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
I think this `for` block and the related containers could be replaced with something like

```java
        // MQ, alignLen, asmID, contigID
        final List<Tuple4<Integer, Integer, String, String>> breakpointAlignmentInformation = StreamSupport.stream(assembledBreakpointsPerAllele._2.spliterator(), false).map(entry ->{
            final BreakpointAlignment breakpointAlignment = entry._2();
            final int assembledBreakpointAlignmentLength =
                    Math.min(breakpointAlignment.region1.referenceInterval.size(),
                            breakpointAlignment.region2.referenceInterval.size()) - breakpointAlignment.region1.overlapOnContig(breakpointAlignment.region2);
            return new Tuple4<>(Math.min(breakpointAlignment.region1.mapqual, breakpointAlignment.region2.mapqual), // mqs
                    assembledBreakpointAlignmentLength, // alignLengths
                    entry._1._1, //assemblyID
                    entry._1._2);//contigID
        }).collect(Collectors.toList());

        final int maxAlignLength = breakpointAlignmentInformation.stream().map(e -> e._2()).max(Comparator.naturalOrder()).orElse(0);
        final int highMqMappings = (int) breakpointAlignmentInformation.stream().map(e -> e._1()).filter(i -> i==60).count();
```

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
And these attributes assignment could be accordingly set this way

```java
.attribute(GATKSVVCFHeaderLines.MAPPING_QUALITIES, breakpointAlignmentInformation.stream().map(e -> e._1()).map(String::valueOf).collect(Collectors.joining(",")))
.attribute(GATKSVVCFHeaderLines.ALIGN_LENGTHS, breakpointAlignmentInformation.stream().map(e -> e._2()).map(String::valueOf).collect(Collectors.joining(",")))
.attribute(GATKSVVCFHeaderLines.ASSEMBLY_IDS, breakpointAlignmentInformation.stream().map(e -> e._3()).collect(Collectors.joining(",")))
.attribute(GATKSVVCFHeaderLines.CONTIG_IDS, breakpointAlignmentInformation.stream().map(e -> e._4()).map(s -> s.replace(" ", "_")).collect(Collectors.joining(",")));
```

SHuang-Broad commented:
Back to you @cwhelan . Sorry for keep bugging you with these.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/BreakpointAlignment.java:
I've refactored these two methods into a single method called `getLeftAndRightBreakpointsOnReferenceLeftAlignedForHomology`. 

I don't think that the word "Parsimony" really is relevant to what the logic is doing here, or that it describes the process -- left aligning for homology is just done for the purpose of getting a canonical breakpoint location for each pair of alignments. 

I left the method `getLeftAlignmentRegion` as is because I think it makes it explicit what the check is in terms of the naming scheme of the other variables.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
I don't really see how adding a stream and collect call makes things any clearer so I'm going to leave this as is. Plus I prefer to leave calls to `collect` on an RDD on their own line so that you can explicitly see the transition from RDD-land to Java-collection-on-the-driver land.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
removed

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
Yes, the filtering sort of gradually moved during refactoring. Changed to `getVariantContextForBreakpointAlleleAlignmentList`.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
I'm not really sure that's easier to read. In a Tuple4 you have to keep track of what four fields are supposed to be without explicitly naming them, which is tough. In this case I think an old fashioned for loop is better.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CallVariantsFromAlignedContigsSpark.java:
same comment as above

cwhelan commented:
@SHuang-Broad refactored some code in response to your comments. how does this look now?

SHuang-Broad commented:
Everything looks fine now. Thanks! 
Please rebase and squash to 1 commit, though.
